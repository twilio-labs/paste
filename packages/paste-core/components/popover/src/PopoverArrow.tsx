import * as React from 'react';
import {
  NonModalDialogArrowPrimitive,
  NonModalDialogArrowPrimitiveProps,
} from '@twilio-paste/non-modal-dialog-primitive';
import {useTheme} from '@twilio-paste/theme';

export type PopoverArrowProps = NonModalDialogArrowPrimitiveProps;

const PopoverArrow: React.FC<PopoverArrowProps> = props => {
  const theme = useTheme();

  return (
    <NonModalDialogArrowPrimitive
      {...props}
      size={theme.fontSizes.fontSize70}
      stroke={theme.borderColors.colorBorderLight}
      fill={theme.backgroundColors.colorBackgroundBody}
    />
  );
};

PopoverArrow.displayName = 'PopoverArrow';
export {PopoverArrow};
