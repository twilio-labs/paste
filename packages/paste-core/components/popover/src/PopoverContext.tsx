import * as React from 'react';
import {NonModalDialogPrimitiveState} from '@twilio-paste/non-modal-dialog-primitive';

const PopoverContext = React.createContext<Partial<NonModalDialogPrimitiveState>>({});

export {PopoverContext};
