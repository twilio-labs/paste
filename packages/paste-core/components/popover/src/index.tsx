export * from './PopoverContainer';
export * from './Popover';
export * from './PopoverButton';
