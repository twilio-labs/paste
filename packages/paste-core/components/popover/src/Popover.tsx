import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box, BoxProps, safelySpreadBoxProps} from '@twilio-paste/box';
import {Button} from '@twilio-paste/button';
import {CloseIcon} from '@twilio-paste/icons/esm/CloseIcon';
import {StyledBase} from '@twilio-paste/theme';
import {NonModalDialogPrimitive} from '@twilio-paste/non-modal-dialog-primitive';
import {PopoverArrow} from './PopoverArrow';
import {PopoverContext} from './PopoverContext';

export interface PopoverProps {
  'aria-label': string;
  children: React.ReactNode;
}

const StyledPopover = React.forwardRef<HTMLDivElement, BoxProps>(({style, ...props}, ref) => {
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      backgroundColor="colorBackgroundBody"
      borderStyle="solid"
      borderWidth="borderWidth10"
      borderColor="colorBorderLighter"
      borderRadius="borderRadius20"
      boxShadow="shadowCard"
      maxWidth="size30"
      zIndex="zIndex80"
      _focus={{outline: 'none'}}
      style={style}
      ref={ref}
    />
  );
});

const Popover = React.forwardRef<HTMLDivElement, PopoverProps>(({children, ...props}, ref) => {
  const popover = React.useContext(PopoverContext);
  return (
    <NonModalDialogPrimitive {...(popover as any)} {...props} as={StyledPopover} ref={ref}>
      {/* import Paste Theme Based Styles due to portal positioning. */}
      <StyledBase>
        <PopoverArrow {...(popover as any)} />
        <Box padding="space50" paddingLeft="space70" paddingRight="space70">
          {children}
          <Box position="absolute" right={8} top={8}>
            <Button
              variant="reset"
              size="reset"
              // @ts-ignore
              // Property 'hide' does not exist on type 'Partial<PopoverState>'
              // But reakit docs suggest using it
              // https://reakit.io/docs/popover/#initial-focus
              onClick={popover.hide}
            >
              <CloseIcon decorative={false} color="colorTextWeak" size="sizeIcon30" title="Close popover" />
            </Button>
          </Box>
        </Box>
      </StyledBase>
    </NonModalDialogPrimitive>
  );
});

if (process.env.NODE_ENV === 'development') {
  Popover.propTypes = {
    'aria-label': PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
  };
}

Popover.displayName = 'Popover';
export {Popover};
