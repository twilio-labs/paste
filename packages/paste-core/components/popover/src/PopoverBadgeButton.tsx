import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Badge} from '@twilio-paste/badge';
import {NonModalDialogDisclosurePrimitive} from '@twilio-paste/non-modal-dialog-primitive';
import type {PopoverBadgeButtonProps, ButtonBadgeProps} from './types';
import {PopoverContext} from './PopoverContext';

// eslint-disable-next-line react/display-name
const ButtonBadge = React.forwardRef<HTMLButtonElement, ButtonBadgeProps>(({children, ...props}, ref) => (
  <Badge {...props} as="button" ref={ref}>
    {children}
  </Badge>
));

const PopoverBadgeButton = React.forwardRef<HTMLButtonElement, PopoverBadgeButtonProps>(
  ({children, element = 'POPOVER_BADGE', ...popoverButtonProps}, ref) => {
    const popover = React.useContext(PopoverContext);

    return (
      <NonModalDialogDisclosurePrimitive
        element={element}
        {...(popover as any)}
        {...popoverButtonProps}
        as={ButtonBadge}
        ref={ref}
      >
        {children}
      </NonModalDialogDisclosurePrimitive>
    );
  }
);

PopoverBadgeButton.propTypes = {
  id: PropTypes.string,
  toggle: PropTypes.func,
  element: PropTypes.string,
};

PopoverBadgeButton.displayName = 'PopoverBadgeButton';
export {PopoverBadgeButton};
