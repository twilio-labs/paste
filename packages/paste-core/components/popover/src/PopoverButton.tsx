import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Button, ButtonProps} from '@twilio-paste/button';
import {NonModalDialogDisclosurePrimitive} from '@twilio-paste/non-modal-dialog-primitive';
import {PopoverContext} from './PopoverContext';

export interface PopoverButtonProps extends ButtonProps {
  id?: string;
  children: React.ReactNode;
  toggle?: () => void;
}

const PopoverButton = React.forwardRef<HTMLButtonElement, PopoverButtonProps>(
  ({children, ...popoverButtonProps}, ref) => {
    const popover = React.useContext(PopoverContext);
    return (
      <NonModalDialogDisclosurePrimitive {...(popover as any)} {...popoverButtonProps} as={Button} ref={ref}>
        {children}
      </NonModalDialogDisclosurePrimitive>
    );
  }
);

if (process.env.NODE_ENV === 'development') {
  PopoverButton.propTypes = {
    id: PropTypes.string,
    children: PropTypes.node.isRequired,
    toggle: PropTypes.func,
  };
}

PopoverButton.displayName = 'PopoverButton';
export {PopoverButton};
