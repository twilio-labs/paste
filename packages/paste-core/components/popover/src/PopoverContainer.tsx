import * as React from 'react';
import {
  useNonModalDialogPrimitiveState,
  NonModalDialogPrimitivePopoverInitialState,
} from '@twilio-paste/non-modal-dialog-primitive';
import {PopoverContext} from './PopoverContext';

export type PopoverContainerProps = NonModalDialogPrimitivePopoverInitialState;

const PopoverContainer: React.FC<PopoverContainerProps> = ({children, placement, gutter, ...initialState}) => {
  const popover = useNonModalDialogPrimitiveState({placement, gutter, modal: true, ...initialState});
  const value = React.useMemo(() => popover, Object.values(popover));
  return <PopoverContext.Provider value={value}>{children}</PopoverContext.Provider>;
};

PopoverContainer.displayName = 'PopoverContainer';
export {PopoverContainer};
