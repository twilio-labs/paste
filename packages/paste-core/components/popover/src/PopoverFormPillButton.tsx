import { FormPill } from "@twilio-paste/form-pill-group";
import { NonModalDialogDisclosurePrimitive } from "@twilio-paste/non-modal-dialog-primitive";
import * as React from "react";

import { PopoverContext } from "./PopoverContext";
import type { PopoverFormPillButtonProps } from "./types";

const PopoverFormPillButton = React.forwardRef<HTMLElement, PopoverFormPillButtonProps>(
  ({ children, element = "POPOVER_FORM_PILL", ...popoverButtonProps }, ref) => {
    const popover = React.useContext(PopoverContext);

    return (
      <NonModalDialogDisclosurePrimitive
        element={element}
        {...(popover as any)}
        {...popoverButtonProps}
        as={FormPill}
        ref={ref}
        onSelect={(e: React.MouseEvent<HTMLButtonElement>) => {
          // @ts-expect-error Property 'toggle' does not exist on type 'Partial<PopoverState>', but it is there as it comes form DialogActions prop.
          popover.toggle();
          // Call the actual onsSelect function passed to the component
          if (popoverButtonProps.onSelect) {
            popoverButtonProps.onSelect(e);
          }
        }}
        baseId={popover.baseId}
      >
        {children}
      </NonModalDialogDisclosurePrimitive>
    );
  },
);

PopoverFormPillButton.displayName = "PopoverFormPillButton";
export { PopoverFormPillButton };
