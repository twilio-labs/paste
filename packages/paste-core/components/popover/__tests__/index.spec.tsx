import * as React from 'react';
import {render, screen} from '@testing-library/react';
import {Theme} from '@twilio-paste/theme';
import axe from '../../../../../.jest/axe-helper';
import {Popover, PopoverContainer, PopoverButton} from '../src';

const PopoverMock: React.FC = () => {
  return (
    <Theme.Provider theme="console">
      <PopoverContainer baseId="popover-example">
        <PopoverButton variant="primary">Open popover</PopoverButton>
        <Popover aria-label="Popover">
          This is the Twilio styled popover that you can use in all your applications.
        </Popover>
      </PopoverContainer>
    </Theme.Provider>
  );
};

const VisiblePopoverMock: React.FC = () => {
  return (
    <Theme.Provider theme="console">
      <PopoverContainer baseId="visible-popover-example" visible>
        <PopoverButton variant="primary" data-testid="open-button">
          Open popover
        </PopoverButton>
        <Popover aria-label="Popover">
          This is the Twilio styled popover that you can use in all your applications.
        </Popover>
      </PopoverContainer>
    </Theme.Provider>
  );
};

describe('Popover', () => {
  describe('Render', () => {
    it('should render', () => {
      const {asFragment} = render(<PopoverMock />);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render a popover button with aria attributes', () => {
      render(<PopoverMock />);
      const renderedPopoverButton = screen.getByRole('button');
      expect(renderedPopoverButton.getAttribute('aria-haspopup')).toEqual('dialog');
      expect(renderedPopoverButton.getAttribute('aria-controls')).toEqual('popover-example');
      expect(renderedPopoverButton.getAttribute('aria-expanded')).toEqual('false');
    });

    it('should render a popover button with expanded aria attributes when popover is visible', () => {
      render(<VisiblePopoverMock />);
      const renderedVisiblePopoverButton = screen.getByTestId('open-button');
      expect(renderedVisiblePopoverButton.getAttribute('aria-expanded')).toEqual('true');
    });

    it('should render a popover', () => {
      render(<PopoverMock />);
      const renderedPopover = screen.getByLabelText('Popover');
      expect(renderedPopover.getAttribute('role')).toEqual('dialog');
    });
  });

  describe('Accessibility', () => {
    it('Should have no accessibility violations', async () => {
      const {container} = render(<PopoverMock />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});
