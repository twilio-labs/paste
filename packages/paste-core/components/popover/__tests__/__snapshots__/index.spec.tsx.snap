// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Render should render 1`] = `
<DocumentFragment>
  .emotion-2 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-1 {
  box-sizing: border-box;
  width: auto;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  display: inline-block;
  border: none;
  outline: none;
  -webkit-transition: background-color 100ms ease-in,box-shadow 100ms ease-in,color 100ms ease-in;
  transition: background-color 100ms ease-in,box-shadow 100ms ease-in,color 100ms ease-in;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  position: relative;
  cursor: pointer;
  color: rgb(255,255,255);
  background-color: rgb(2,122,197);
  box-shadow: 0 0 0 1px #027ac5;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  border-radius: 4px;
  font-size: 0.875rem;
  line-height: 1.25rem;
}

.emotion-1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: rgb(255,255,255);
  background-color: rgb(3,44,94);
  box-shadow: 0 0 0 1px #032c5e;
}

.emotion-1:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: 0 0 0 4px rgba(2,122,197,0.8);
  color: rgb(255,255,255);
  background-color: rgb(3,44,94);
}

.emotion-1:active,
.emotion-1[data-active=true] {
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: 0 0 0 1px #032c5e;
  color: rgb(255,255,255);
  background-color: rgb(0,62,130);
}

.emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  opacity: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div
    class="emotion-2"
  >
    <button
      aria-busy="false"
      aria-controls="popover-example"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="emotion-1"
      color="colorTextInverse"
      cursor="pointer"
      display="inline-block"
      font-family="fontFamilyText"
      font-size="fontSize30"
      font-weight="fontWeightSemibold"
      text-decoration="none"
      type="button"
      width="auto"
    >
      <span
        aria-hidden="false"
        class="emotion-0"
        display="flex"
        opacity="1"
        text-decoration="inherit"
      >
        Open popover
      </span>
    </button>
  </div>
</DocumentFragment>
`;
